libs/models/datapoints.py

```diff
@@ -0,0 +1,400 @@
+            logger.error(
+                f"Error creating production breakdown datapoint {datetime}: {e}"
+            )
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "datetime": self.datetime,
+            "zoneKey": self.zoneKey,
+            "production": self.production.dict(),
+            "storage": self.storage.dict() if self.storage else None,
+            "source": self.source,
+        }
+
+
+class Consumption(Datapoint):
+    consumption: float
+
+    @validator("consumption")
+    def _validate_consumption(cls, v: float):
+        if v < 0:
+            raise ValueError(f"Consumption cannot be negative: {v}")
+        if v > 500000:
+            raise ValueError(f"Consumption is implausibly high, above 500GW: {v}")
+
+    @staticmethod
+    def create(
+        logger: Logger,
+        zone_key: ZoneKey,
+        datetime: datetime,
+        source: dict,
+        consumption: float,
+        forecasted: bool = False,
+    ) -> Optional["Consumption"]:
+        try:
+            return Consumption(
+                zoneKey=zone_key,
+                datetime=datetime,
+                source=source,
+                consumption=consumption,
+                forecasted=forecasted,
+            )
+        except ValueError as e:
+            logger.error(
+                f"Error creating consumption datapoint {datetime}: {e}",
+                extra={
+                    "zone_key": zone_key,
+                    "datetime": datetime,
+                    "kind": "consumption",
+                },
+            )
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "datetime": self.datetime,
+            "zoneKey": self.zoneKey,
+            "consumption": self.consumption,
+            "source": self.source,
+        }
+
+
+class Price(Datapoint):
+    price: float
+    currency: str
+
+    @validator("currency")
+    def _validate_currency(cls, v: str):
+        if v not in VALID_CURRENCIES:
+            raise ValueError(f"Unknown currency: {v}")
+        return v
+
+    @staticmethod
+    def create(
+        logger: Logger,
+        zone_key: ZoneKey,
+        datetime: datetime,
+        source: dict,
+        price: float,
+        currency: str,
+        forecasted: bool = False,
+    ) -> Optional["Price"]:
+        try:
+            return Price(
+                zoneKey=zone_key,
+                datetime=datetime,
+                source=source,
+                price=price,
+                currency=currency,
+                forecasted=forecasted,
+            )
+        except ValueError as e:
+            logger.error(f"Error creating price datapoint {datetime}: {e}")
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "datetime": self.datetime,
+            "zoneKey": self.zoneKey,
+            "currency": self.currency,
+            "price": self.price,
+            "source": self.source,
+        }
+
+
+class DatapointBatch(ABC):
```

Comment:
Could we rename this to something other than batch? I think maybe just naming them as list would make the most sense? Especially since we use the same syntax as list with append.

Or perhaps set as we expect these to be unique?

I think it's a good idea to leverage the already existing naming conventions for clarity, especially since we have a lot of open source contributors.

Reply:
Renamed to list